File: py-rosalind/revc.py

def complement(stream):
    for line in stream:
        for c in line:
            if c == "A":
                yield "T"
            elif c == "C":
                yield "G"
            elif c == "G":
                yield "C"
            elif c == "T":
                yield "A"

def revc(stream):
    return "".join(complement(stream))[::-1]

File: rosalind/rust-rosalind/src/revc.rs
use std::io::{BufReader, BufRead};

pub fn complement(c: char) -> Option<char> {
    match c {
        'A' => Some('T'),
        'C' => Some('G'),
        'G' => Some('C'),
        'T' => Some('A'),
        _ => None
    }
}

pub fn revc(reader: BufReader<std::fs::File>) -> String {
    let mut ns = Vec::new();
    for line in reader.lines() {
        let line = line.expect("Could not read the next line");
        for c in line.chars() {
            if let Some(complement) = complement(c) {
                ns.push(complement)
            }
        }
    }

    ns.iter().rev().collect()
}